File: bap-java/src/algorithm/branchandprice/BapNode.java
package algorithm.branchandprice;

import model.Solution;

import java.util.List;
import java.util.Map;

public abstract class BapNode implements Comparable<BapNode> {

    private boolean isIntegral;
    private int iteration;
    private double nodeObjectiveVal;
    private Solution solution;
    private boolean isLPFeasible;
    private BapNode parent;
    private Arc branchArc;
    private int nodeID;
    private Map<Record, Double> distMap;  // to store the distance map for inheritance
    private List<Integer> infeasiblePathIndices;

    /**
     * Column Generation process.
     *
     * @param rmp Restricted Master Problem.
     * @param sp Pricing sub Problem.
     */
    public void columnGeneration(RestrictedMasterProblem rmp, PricingProblem sp) {

    }

    /**
     * Update the feasible path indices in rmp after branching.
     *
     * @param rmp Restricted Master Problem
     */
    protected abstract void updateInfeasiblePath(RestrictedMasterProblem rmp);

    /**
     * Resolve distance map on graph after branching.
     */
    protected abstract void resolveGraph();

    @Override
    public int compareTo(BapNode bapNode) {
        if (nodeObjectiveVal <= bapNode.nodeObjectiveVal) {
            return -1;
        } else {
            return 1;
        }
    }
}

File: bap-python/src/algorithm/branchandprice/BapNode.py
# **********************************************************
# * Author        : XXXXXX
# * Email         : X@XXX.com
# * Create time   : 2022/XX/XX X:XX AM
# * Filename      : BapNode
# * Description   :
# **********************************************************
from src.algorithm.branchandprice.PricingProblem import PricingProblem
from src.algorithm.branchandprice.RestrictedMasterProblem import RestrictedMasterProblem
from src.utils.Util import Solution


class BapNode:

    """ BapNode class for branch-and-bound procedure in Branch-and-Price algorithm. """

    def __init__(self):
        self.is_integral: bool = False
        self.iteration: int = 0
        self.node_obj: float = float('inf')
        self.solution: Solution = None
        self.is_LP_feasible: bool = True
        self.parent: BapNode = None
        self.branch_arc: tuple = None
        self.node_id: int = 1
        self.dist_map: dict = None
        self.infeasible_path_indices: list = None

    def __lt__(self, other):
        pass

    def column_generation(self, rmp: RestrictedMasterProblem, sp: PricingProblem) -> None:
        """
        Column Generation process.

        :param rmp: restricted master problem
        :param sp: pricing sub problem
        """
        pass

    def _update_feasible_path(self, rmp: RestrictedMasterProblem) -> None:
        """
        Update the feasible path indices in rmp after branching.

        :param rmp: restricted master problem
        """
        pass

    def _resolve_graph(self) -> None:
        """ Resolve distance map on graph after branching. """
        pass
